File: sources/scala/tools/util/SourceReader.java
/*     ____ ____  ____ ____  ______                                     *\
**    / __// __ \/ __// __ \/ ____/    SOcos COmpiles Scala             **
**  __\_ \/ /_/ / /__/ /_/ /\_ \       (c) 2002, LAMP/EPFL              **
** /_____/\____/\___/\____/____/                                        **
\*                                                                      */

// $Id$

package scala.tools.util;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.CharBuffer;
import java.nio.channels.FileChannel;
import java.nio.channels.ReadableByteChannel;
import java.nio.charset.CharsetDecoder;
import java.nio.charset.CoderResult;

/** This class implements methods to read and decode source files. */
public class SourceReader {

    //########################################################################
    // Private Fields

    /** The decoder used to transform bytes into characters */
    private final CharsetDecoder decoder;

    /** The input byte buffer (small enough to fit in cache) */
    private final ByteBuffer bytes;

    /** The output character buffer */
    private CharBuffer chars;

    //########################################################################
    // Public Constructors

    /** Initializes this instance with the specified decoder. */
    public SourceReader(CharsetDecoder decoder) {
        this.decoder = decoder;
        this.bytes = ByteBuffer.allocate(0x4000);
        this.chars = CharBuffer.allocate(0x4000);
    }

    //########################################################################
    // Public Methods

    /** Reads the file with the specified name. */
    public char[] read(String filename) throws IOException {
        return read(new File(filename));
    }

    /** Reads the specified file. */
    public char[] read(File file) throws IOException {
        FileChannel channel = new FileInputStream(file).getChannel();
        try {
            return read(channel);
        } finally {
            channel.close();
        }
    }

    /** Reads the specified file. */
    public char[] read(AbstractFile file) throws IOException {
        CharsetDecoder decoder = this.decoder.reset();
        ByteBuffer bytes = ByteBuffer.wrap(file.read());
        CharBuffer chars = this.chars; chars.clear();
        return terminate(flush(decoder, decode(decoder, bytes, chars, true)));
    }

    /** Reads the specified byte channel. */
    public char[] read(ReadableByteChannel input) throws IOException {
        CharsetDecoder decoder = this.decoder.reset();
        ByteBuffer bytes = this.bytes; bytes.clear();
        CharBuffer chars = this.chars; chars.clear();
        for (boolean endOfInput = false; !endOfInput; ) {
            endOfInput = input.read(bytes) < 0;
            bytes.flip();
            chars = decode(decoder, bytes, chars, endOfInput);
        }
        return terminate(flush(decoder, chars));
    }

    //########################################################################
    // Private Methods

    /**
     * Sets the specified char buffer as the new output buffer and
     * reads and returns its content.
     */
    private char[] terminate(CharBuffer chars) {
        char[] result = new char[chars.length()];
        chars.get(result);
        this.chars = chars;
        return result;
    }

    //########################################################################
    // Private Functions

    /**
     * Decodes the content of the specified byte buffer with the
     * specified decoder into the specified char buffer, allocating
     * bigger ones if necessary, then compacts the byte buffer and
     * returns the last allocated char buffer. The "endOfInput"
     * argument indicates whether the byte buffer contains the last
     * chunk of the input file.
     */
    private static CharBuffer decode(CharsetDecoder decoder, ByteBuffer bytes,
        CharBuffer chars, boolean endOfInput) throws IOException
    {
        while (true) {
            CoderResult result = decoder.decode(bytes, chars, endOfInput);
            if (result.isUnderflow()) break;
            if (result.isError()) throw new IOException(result.toString());
            assert result.isOverflow();
            chars = increaseCapacity(chars);
        }
        bytes.compact();
        return chars;
    }

    /**
     * Flushes the specified decoder into the specified char buffer,
     * allocating bigger ones if necessary and then flips and returns
     * the last allocated char buffer.
     */
    private static CharBuffer flush(CharsetDecoder decoder, CharBuffer chars)
        throws IOException
    {
        while (true) {
            CoderResult result = decoder.flush(chars);
            if (result.isUnderflow()) break;
            if (result.isError()) throw new IOException(result.toString());
            assert result.isOverflow();
            chars = increaseCapacity(chars);
        }
        chars.flip();
        return chars;
    }

    /**
     * Flips the specified buffer and returns a new one with the same
     * content but with an increased capacity.
     */
    private static CharBuffer increaseCapacity(CharBuffer buffer) {
        buffer.flip();
        int capacity = 2 * buffer.capacity();
        return CharBuffer.allocate(capacity).put(buffer);
    }

    //########################################################################

}

File: src/compiler/scala/tools/nsc/io/SourceReader.scala
/*
 * Scala (https://www.scala-lang.org)
 *
 * Copyright EPFL and Lightbend, Inc. dba Akka
 *
 * Licensed under Apache License 2.0
 * (http://www.apache.org/licenses/LICENSE-2.0).
 *
 * See the NOTICE file distributed with this work for
 * additional information regarding copyright ownership.
 */

package scala.tools.nsc
package io

import java.io.{FileInputStream, IOException}
import java.nio.{ByteBuffer, CharBuffer}
import java.nio.channels.{Channels, ClosedByInterruptException, ReadableByteChannel}
import java.nio.charset.{CharsetDecoder, CoderResult}
import scala.annotation.tailrec
import scala.reflect.internal.Reporter

/** This class implements methods to read and decode source files. */
class SourceReader(decoder: CharsetDecoder, reporter: Reporter) {

  import SourceReader.{decode, flush}

  //########################################################################
  // Private Fields

  /** The input byte buffer (small enough to fit in cache) */
  private val bytes: ByteBuffer = ByteBuffer.allocate(0x4000)

  /** The output character buffer */
  private var chars: CharBuffer = CharBuffer.allocate(0x4000)

  private def reportEncodingError(filename: String, e: Exception) = {
    val advice = "Please try specifying another one using the -encoding option"
    reporter.error(scala.reflect.internal.util.NoPosition,
      s"IO error while decoding $filename with ${decoder.charset()}: ${e.getMessage}\n$advice")
  }

  /** Reads the specified file. */
  def read(file: JFile): Array[Char] = {
    val c = new FileInputStream(file).getChannel

    try read(c)
    catch {
      case ex: InterruptedException => throw ex
      case _: ClosedByInterruptException => throw new InterruptedException
      case e: Exception => reportEncodingError("" + file, e) ; Array()
    }
    finally c.close()
  }

  /** Reads the specified file.
   */
  def read(file: AbstractFile): Array[Char] = {
    try file match {
      case p: PlainFile        => read(p.file)
      case z: ZipArchive#Entry => val c = Channels.newChannel(z.input); try read(c) finally c.close()
      case vf: VirtualFile     => read(ByteBuffer.wrap(vf.unsafeToByteArray))
      case f                   => val c = Channels.newChannel(f.input); try read(c) finally c.close()
    }
    catch {
      case ex: InterruptedException => throw ex
      case _: ClosedByInterruptException => throw new InterruptedException
      case e: Exception => reportEncodingError("" + file, e) ; Array()
    }
  }

  /** Reads the specified byte channel. */
  protected def read(input: ReadableByteChannel): Array[Char] = {
    val decoder: CharsetDecoder = this.decoder.reset()
    val bytes: ByteBuffer       = this.bytes; bytes.clear()
    var chars: CharBuffer       = this.chars; chars.clear()
    var endOfInput              = false

    while (!endOfInput ) {
      endOfInput = input.read(bytes) < 0
      bytes.flip()
      chars = decode(decoder, bytes, chars, endOfInput)
    }
    terminate(flush(decoder, chars))
  }

  /** Reads the specified byte buffer. */
  protected def read(bytes: ByteBuffer): Array[Char] = {
    val decoder: CharsetDecoder = this.decoder.reset()
    val chars: CharBuffer = this.chars; chars.clear()
    terminate(flush(decoder, decode(decoder, bytes, chars, endOfInput = true)))
  }

  //########################################################################
  // Private Methods

  /**
   * Sets the specified char buffer as the new output buffer and
   * reads and returns its content.
   */
  private def terminate(chars: CharBuffer): Array[Char] = {
    val result = new Array[Char](chars.length())
    chars.get(result)
    this.chars = chars
    result
  }

}

object SourceReader {

  /**
   * Decodes the content of the specified byte buffer with the
   * specified decoder into the specified char buffer, allocating
   * bigger ones if necessary, then compacts the byte buffer and
   * returns the last allocated char buffer. The "endOfInput"
   * argument indicates whether the byte buffer contains the last
   * chunk of the input file.
   */
  @tailrec
  def decode(decoder: CharsetDecoder, bytes: ByteBuffer, chars: CharBuffer,
             endOfInput: Boolean): CharBuffer =
  {
    val result: CoderResult = decoder.decode(bytes, chars, endOfInput)
    if (result.isUnderflow()) {
      bytes.compact()
      chars
    } else {
      if (result.isError()) throw new IOException(result.toString())
      assert(result.isOverflow())
      decode(decoder, bytes, increaseCapacity(chars), endOfInput)
    }
  }

  /**
   * Flushes the specified decoder into the specified char buffer,
   * allocating bigger ones if necessary and then flips and returns
   * the last allocated char buffer.
   */
  @tailrec
  def flush(decoder: CharsetDecoder, chars: CharBuffer): CharBuffer = {
    val result: CoderResult = decoder.flush(chars)
    if (result.isUnderflow()) {
      chars.flip()
      chars
    } else {
      if (result.isError()) throw new IOException(result.toString())
      assert(result.isOverflow())
      flush(decoder, increaseCapacity(chars))
    }
  }

  /**
   * Flips the specified buffer and returns a new one with the same
   * content but with an increased capacity.
   */
  private def increaseCapacity(buffer: CharBuffer): CharBuffer = {
    buffer.flip()
    val capacity = 2 * buffer.capacity()
    CharBuffer.allocate(capacity).put(buffer)
  }

}
